### PR TITLE
global attention pooling added

### PR DESCRIPTION
**Pull request by sdkaraban**

### CODE REVIEW

File: src/pooling.py [hunk 1]
```diff
@@ -0,0 +1,30 @@
+import torch
+import torch.nn as nn
+
+
+class GlobalAttentionPooling(nn.Module):
+    def __init__(self, input_dim, hidden_dim=None):  # No need for hidden_dim for global pooling
```
Comment (ilsenatorov) [inline src/pooling.py:6]: if you don't use hidden_dim, maybe just remove it?

**Review by ilsenatorov — CHANGES_REQUESTED** [review]

**Comment by devium335** [comment]
Why is beartype being disabled?

File: src/pooling.py [hunk 2]
```diff
@@ -0,0 +1,30 @@
...
+
+class GlobalAttentionPooling(nn.Module):
+    def __init__(self, input_dim):
+        super(GlobalAttentionPooling, self).__init__()
+
+        self.linear_key = nn.Linear(input_dim, 1)  # Project to a single dimension
+        self.linear_query = nn.Linear(input_dim, 1)
+        self.softmax = nn.Softmax(dim=-1)
+
+    def forward(self, x):
```
Comment (devium335) [inline src/pooling.py:13]: Add type hinting

File: src/pooling.py [hunk 3]
```diff
@@ -0,0 +1,30 @@
...
+        attention_weights = self.softmax(attention_scores)  # (batch_size, seq_len, seq_len)
+
+        # Global weighted sum (sum over sequence length)
+        pooled_output = torch.matmul(attention_weights, x).sum(dim=1)  # (batch_size, embedding_dim)
+
+        return pooled_output
+
+
+class MeanPooling(nn.Module):
+    def forward(x):
```
Comment (devium335) [inline src/pooling.py:29]: Type hinting

File: train.py [hunk 4]
```diff
@@ -68,9 +69,15 @@ def train(
 
     train, val, test = random_split(dataset, [0.7, 0.2, 0.1])
 
-    train_dataloader = DataLoader(train, batch_size=batch_size, shuffle=True)
-    validation_dataloader = DataLoader(val, batch_size=batch_size)
-    test_dataloader = DataLoader(test, batch_size=batch_size)
+    def collate_fn(batch):
```
Comment (devium335) [inline train.py:72]: Type hinting

**Review by devium335 — CHANGES_REQUESTED** [review]
Please add type hinting and results from testing code

**Pull request closed without merging**